sdk/apimanagement/api-management-custom-widgets-tools/src/utils.ts

```diff
@@ -0,0 +1,170 @@
+ *
+ * @param valuesDefault - object with your default values to use, just import valuesDefault object from values.ts folder
+ */
+export function getEditorData<TValues extends TValuesCommon>(
+  valuesDefault: TValues
+): TEditorData<TValues> {
+  return getEditorDataPure(valuesDefault, new URLSearchParams(self.location.search));
+}
+
+export function getEditorValuesPure<TValues extends TValuesCommon>(
+  valuesDefault: TValues,
+  urlSearchParams: URLSearchParams
+): TValues {
+  return getEditorDataPure(valuesDefault, urlSearchParams).values;
+}
+
+/**
+ * Function to get values you've set in the admin editor window.
+ *
+ * @param valuesDefault - object with your default values to use, just import valuesDefault object from values.ts folder
+ */
+export function getEditorValues<TValues extends TValuesCommon>(valuesDefault: TValues): TValues {
+  return getEditorValuesPure(valuesDefault, new URLSearchParams(self.location.search));
+}
+
+/**
+ * Type of the onChange function.
+ */
+export type TOnChange<TValues extends TValuesCommon> = (values: Partial<TValues>) => void;
+
+/**
+ * The onChange function itself with 'origin' provided as a param.
+ *
+ * @param origin - web content's origin (URL) of your Dev Portal to send changes to
+ * @param instanceId - ID of this particular instance of the widget
+ * @param values - values that changed
+ */
+export function onChangeWithOrigin<TValues extends TValuesCommon>(
+  origin: TEditorData<TValues>["origin"],
+  instanceId: TEditorData<TValues>["instanceId"],
+  values: TValues
+): void {
+  Object.entries(values).forEach(([key, value]) => {
+    self.parent.postMessage({ [ON_CHANGE_MESSAGE_KEY]: { key, value, instanceId } }, origin);
+  });
+}
+
+/**
+ * Build onChange function, which you can use, to send changed data from the editor.
+ *
+ * @param valuesDefault - object with your default values to use, just import valuesDefault object from values.ts folder
+ */
+export function buildOnChange<TValues extends TValuesCommon>(
+  valuesDefault: TValues
+): TOnChange<TValues> {
+  const { origin, instanceId } = getEditorData(valuesDefault);
+  return (values: Partial<TValues>) => onChangeWithOrigin(origin, instanceId, values);
+}
+
+/**
+ * Possible target modules
+ * "app" for main application which is embedded in your Dev Portal
+ * "editor" for form in admin panel
+ */
+export type TTargetModule = "app" | "editor";
+/**
+ * Secrets needed for communication with Dev Portal back-end
+ */
+export type TSecrets = { token: string; userId: string };
+/**
+ * Request secrets - token & userId, from the Dev portal parent window.
+ *
+ * @param targetModule - is the function invoke from the main "app" window or the admin "editor"?
+ * @param targetOrigin - web content's origin (URL) of your Dev Portal to send changes to
+ * @param instanceId - ID of this particular instance of the widget
+ * @param environment - what environment is it running on
+ */
+export const askForSecrets = async (
```

Comment:
I'd recommend to be consistent on how you define functions, I see a mix of arrow key constants and function